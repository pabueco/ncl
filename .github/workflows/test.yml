name: Run Tests

on:
  push:
    branches:
      - main

jobs:
  test:
    name: test
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v1

      - run: bun install --frozen-lockfile
      - run: bun test
