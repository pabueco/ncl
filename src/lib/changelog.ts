import { marked } from "marked";
import type { Release, ReleaseWithTokens, VersionParams } from "../types";
import { coerceToSemVer, versionSatisfiesParams } from "./version";

export async function loadChangelogFile(url: string): Promise<string> {
  const res = await fetch(url);
  if (!res.ok) {
    return "";
  }

  return await res.text();
}

export async function parseReleasesFromChangelog(
  source: string,
  versionParams: VersionParams
): Promise<Release[]> {
  if (!source) {
    return [];
  }

  const tokens = marked.lexer(source);

  const releases: ReleaseWithTokens[] = [];

  let currentRelease: ReleaseWithTokens | null = null;
  for (const token of tokens) {
    if (token.type === "heading" && [1, 2, 3].includes(token.depth)) {
      const text = token.text.trim();
      const version = coerceToSemVer(text);

      // If we can't find a version, treat it like any other token.
      if (!version) {
        currentRelease?.content.push(token);
        continue;
      }

      if (currentRelease) {
        releases.unshift(currentRelease);
      }

      if (!versionSatisfiesParams(version, versionParams)) {
        currentRelease = null;
        continue;
      }

      currentRelease = {
        version: version.toString(),
        content: [],
      };
    }

    currentRelease?.content.push(token);
  }

  if (currentRelease && currentRelease.content.length > 0) {
    releases.unshift(currentRelease);
  }

  return releases;
}

export function isChangelogUrl(url: string) {
  try {
    new URL(url);
  } catch {
    return false;
  }

  // Ends with "changelog" or "changelog.md" or "changelog-<anything>.md" (like "changelog-1.0.md")
  const regex = /changelog(-.*)?(\.md)?$/i;
  return regex.test(url);
}
