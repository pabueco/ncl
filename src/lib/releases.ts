import { $ } from "bun";
import { GITHUB_PAGE_LIMIT } from "../constants";
import type { RawGitHubRelease, Release, VersionParams } from "../types";
import { debug } from "../utils";
import { findValidVersionInStrings, versionSatisfiesParams } from "./version";
import { format } from "date-fns";
import type { marked as markedType } from "marked";

export async function loadGitHubReleases(
  repoName: string,
  versionParams: VersionParams
): Promise<Release[]> {
  const rawReleases: any[] = [];
  let page = 1;
  let hasFoundStart = false;

  while (page <= GITHUB_PAGE_LIMIT) {
    debug(`Fetching page ${page}...`);

    // const releases =
    //   await $`gh release list --repo ${repoName} --json name,tagName,isLatest --exclude-pre-releases --exclude-drafts --order desc --limit 100`.quiet();
    // We have to use the API because the CLI does not support paging/offsets.
    const releases =
      await $`gh api -H 'Accept: application/vnd.github+json' -H 'X-GitHub-Api-Version: 2022-11-28' repos/${repoName}/releases -X GET -f per_page=100 -f page=${page}`.quiet();

    const releasesJson: RawGitHubRelease[] = releases.json();

    if (!releasesJson.length) {
      debug(`No more releases found, stopping.`);
      break;
    }

    if (!hasFoundStart) {
      hasFoundStart = releasesJson.some((r) => {
        const version = findValidVersionInStrings([r.tag_name, r.name]);
        return versionSatisfiesParams(version, versionParams);
      });
    }

    rawReleases.push(...releasesJson);

    // This is not working as expected when the last version is not satisfying the params, but the ones on the previous page are.
    // This can happen with pre-releases for example. Fetching everything is probably the best option.
    // TODO: Find a better way to not have to fetch all releases.
    // const lastRelease = releasesJson[releasesJson.length - 1];
    // const lastVersion = findValidVersionInStrings([
    //   lastRelease.tag_name,
    //   lastRelease.name,
    // ]);
    // if (hasFoundStart && !versionSatisfiesParams(lastVersion, versionParams)) {
    //   debug(`Reached version ${lastVersion}, stopping.`);
    //   break;
    // }

    page++;
  }

  return rawReleases
    .toReversed()
    .map((r) => ({
      version:
        findValidVersionInStrings([r.tag_name, r.name])?.toString() || "",
      content: r.body as string,
      url: r.html_url,
      date: format(new Date(r.published_at), "yyyy-MM-dd"),
    }))
    .filter((r) => {
      return versionSatisfiesParams(r.version, versionParams);
    });
}

export async function renderRelease(
  release: Release,
  marked: typeof markedType
): Promise<string> {
  return typeof release.content === "string"
    ? await marked(release.content)
    : marked.parser(release.content);
}
