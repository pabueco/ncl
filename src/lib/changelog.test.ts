import { describe, it, expect, mock } from "bun:test";
import { isChangelogUrl, parseReleasesFromChangelog } from "./changelog";
import { renderRelease } from "./releases";
import { marked } from "marked";

describe("isChangelogUrl", () => {
  it("returns true for changelog URL", () => {
    expect(isChangelogUrl("https://example.com/changelog.md")).toBe(true);
    expect(isChangelogUrl("https://example.com/CHANGELOG.md")).toBe(true);
    expect(isChangelogUrl("https://example.com/changelog-1.md")).toBe(true);
    expect(isChangelogUrl("https://example.com/changelog-3.1.md")).toBe(true);
    expect(isChangelogUrl("https://example.com/CHANGELOG-3.1.md")).toBe(true);
  });
});

describe("parseReleasesFromChangelog", () => {
  it("parses changelog with valid keepachangelog format", async () => {
    const releases = await parseReleasesFromChangelog(
      `
# Changelog

## [1.0.0] - 2024-04-18

### Added
- Foo feature
- Bar feature

### Changed
- Baz feature

## [0.1.0] - 2024-03-01

### Added
- New feature
- Another feature

### Fixed
- Bug fix

## [0.0.1] - 2023-12-30

### Removed
- Old feature
`,
      () => true
    );

    expect(releases).toHaveLength(3);
    expect(releases[0].version).toBe("0.0.1");
    expect(releases[1].version).toBe("0.1.0");
    expect(releases[2].version).toBe("1.0.0");

    expect(await renderRelease(releases[0], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[1], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[2], marked)).toMatchSnapshot();
  });

  it("parses changelog with different format", async () => {
    const releases = await parseReleasesFromChangelog(
      `
# [1.0.0] - 2024-04-18
- Added foo feature
- Fixed bar feature

# [0.1.0] - 2024-03-01
- Added new feature
- Added another feature

# [0.0.1] - 2023-12-30
- Old feature removed
`,
      () => true
    );

    expect(releases).toHaveLength(3);
    expect(releases[0].version).toBe("0.0.1");
    expect(releases[1].version).toBe("0.1.0");
    expect(releases[2].version).toBe("1.0.0");

    expect(await renderRelease(releases[0], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[1], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[2], marked)).toMatchSnapshot();
  });

  it("parses changelog with inconsistent format", async () => {
    const releases = await parseReleasesFromChangelog(
      `
# Changelog

Hello this is my changelog.

## [1.0.0] - 2024-04-18

# Added
- Added foo feature
- Fixed bar feature

### Changed
- Baz feature

### [0.1.0] - 2024-03-01
- Added new feature
- Added another feature

# [0.0.1] - 2023-12-30

### Removed
- Old feature removed
`,
      () => true
    );

    expect(releases).toHaveLength(3);
    expect(releases[0].version).toBe("0.0.1");
    expect(releases[1].version).toBe("0.1.0");
    expect(releases[2].version).toBe("1.0.0");

    expect(await renderRelease(releases[0], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[1], marked)).toMatchSnapshot();
    expect(await renderRelease(releases[2], marked)).toMatchSnapshot();
  });
});
