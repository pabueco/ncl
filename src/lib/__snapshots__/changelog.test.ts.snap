// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parseReleasesFromChangelog parses changelog with valid keepachangelog format 1`] = `
"<h2>[0.0.1] - 2023-12-30</h2>
<h3>Removed</h3>
<ul>
<li>Old feature</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with valid keepachangelog format 2`] = `
"<h2>[0.1.0] - 2024-03-01</h2>
<h3>Added</h3>
<ul>
<li>New feature</li>
<li>Another feature</li>
</ul>
<h3>Fixed</h3>
<ul>
<li>Bug fix</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with valid keepachangelog format 3`] = `
"<h2>[1.0.0] - 2024-04-18</h2>
<h3>Added</h3>
<ul>
<li>Foo feature</li>
<li>Bar feature</li>
</ul>
<h3>Changed</h3>
<ul>
<li>Baz feature</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with different format 1`] = `
"<h1>[0.0.1] - 2023-12-30</h1>
<ul>
<li>Old feature removed</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with different format 2`] = `
"<h1>[0.1.0] - 2024-03-01</h1>
<ul>
<li>Added new feature</li>
<li>Added another feature</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with different format 3`] = `
"<h1>[1.0.0] - 2024-04-18</h1>
<ul>
<li>Added foo feature</li>
<li>Fixed bar feature</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with inconsistent format 1`] = `
"<h1>[0.0.1] - 2023-12-30</h1>
<h3>Removed</h3>
<ul>
<li>Old feature removed</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with inconsistent format 2`] = `
"<h3>[0.1.0] - 2024-03-01</h3>
<ul>
<li>Added new feature</li>
<li>Added another feature</li>
</ul>
"
`;

exports[`parseReleasesFromChangelog parses changelog with inconsistent format 3`] = `
"<h2>[1.0.0] - 2024-04-18</h2>
<h1>Added</h1>
<ul>
<li>Added foo feature</li>
<li>Fixed bar feature</li>
</ul>
<h3>Changed</h3>
<ul>
<li>Baz feature</li>
</ul>
"
`;
